cedar-policy-validator/src/cedar_schema/err.rs

```diff
@@ -343,86 +360,305 @@ impl Diagnostic for ToJsonSchemaErrors {
+        Self::DuplicatePrincipalOrResource(DuplicatePrincipalOrResource {
+            name: name.to_smolstr(),
+            kind: PR::Resource,
+            loc1,
+            loc2,
+        })
+    }
+
+    pub(crate) fn no_principal(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::NoPrincipalOrResource(NoPrincipalOrResource {
+            kind: PR::Principal,
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn no_resource(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::NoPrincipalOrResource(NoPrincipalOrResource {
+            kind: PR::Resource,
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn reserved_name(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::ReservedName(ReservedName {
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn reserved_keyword(keyword: impl ToSmolStr, loc: Loc) -> Self {
+        Self::ReservedSchemaKeyword(ReservedSchemaKeyword {
+            keyword: keyword.to_smolstr(),
+            loc,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+#[error("this uses a reserved schema keyword: `{keyword}`")]
+pub struct ReservedSchemaKeyword {
+    keyword: SmolStr,
+    loc: Loc,
 }
 
-impl Diagnostic for ToJsonSchemaError {
+impl Diagnostic for ReservedSchemaKeyword {
     fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
-        match self {
-            ToJsonSchemaError::DuplicateDeclarations { loc1, loc2, .. }
-            | ToJsonSchemaError::DuplicateContext { loc1, loc2 }
-            | ToJsonSchemaError::DuplicatePrincipalOrResource { loc1, loc2, .. }
-            | ToJsonSchemaError::DuplicateKeys { loc1, loc2, .. } => Some(Box::new(
-                vec![
-                    LabeledSpan::underline(loc1.span),
-                    LabeledSpan::underline(loc2.span),
-                ]
-                .into_iter(),
-            )),
-            ToJsonSchemaError::DuplicateNameSpaces { loc1, loc2, .. } => {
-                Some(Box::new([loc1, loc2].into_iter().filter_map(|loc| {
-                    Some(LabeledSpan::underline(loc.as_ref()?.span))
-                })))
-            }
-            ToJsonSchemaError::UnknownTypeName(node)
-            | ToJsonSchemaError::ReservedName(node)
-            | ToJsonSchemaError::ReservedSchemaKeyword(node) => Some(Box::new(std::iter::once(
-                LabeledSpan::underline(node.loc.span),
-            ))),
-            ToJsonSchemaError::NoPrincipalOrResource { loc, .. } => {
-                Some(Box::new(std::iter::once(LabeledSpan::underline(loc.span))))
-            }
-        }
+        underline_spans(once(self.loc.span))
+    }
+
+    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
+        Some(&self.loc.src as &dyn miette::SourceCode)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
```

Comment:
Add a help message like "Names containing `__cedar` (e.g, `__cedar`, `__cedar::A`, `A::__cedar`, and `A::__cedar::B`) are reserved"?